sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala

```diff
@@ -376,6 +376,17 @@ object Decimal {
 
   def apply(value: String): Decimal = new Decimal().set(BigDecimal(value))
 
+  // This is used for RowEncoder to handle Decimal inside external row.
+  def apply(value: Any): Decimal = {
+    value match {
+      case j: java.math.BigDecimal => apply(j)
+      case d: Decimal => d
+      case i: java.lang.Integer => apply(i.asInstanceOf[Int])
+      case l: java.lang.Long => apply(l.asInstanceOf[Long])
+      case d: java.lang.Double => apply(d.asInstanceOf[Double])
```

Comment:
This will also allow int, long and double exist in a decimal type field of external row, is it what we want?  cc @rxin